samples/bluetooth/hci_spi/src/main.c

```diff
@@ -0,0 +1,295 @@
+#else
+#define TX_BUF_COUNT 6
+#endif
+
+NET_BUF_POOL_DEFINE(acl_tx_pool, TX_BUF_COUNT, BT_BUF_ACL_SIZE,
+		    BT_BUF_USER_DATA_MIN, NULL);
+
+static struct device *spi_hci_dev;
+static struct device *gpio_dev;
+static BT_STACK_NOINIT(bt_tx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread bt_tx_thread_data;
+
+static K_SEM_DEFINE(sem_spi_rx, 0, 1);
+static K_SEM_DEFINE(sem_spi_tx, 0, 1);
+
+static inline int spi_send(struct net_buf *buf)
+{
+	u8_t header_master[5] = { 0 };
+	u8_t header_slave[5] = { READY_NOW, SANITY_CHECK,
+				 0x00, 0x00, 0x00 };
+	u8_t ret;
+
+	SYS_LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf),
+		    buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_IN:
+		net_buf_push_u8(buf, HCI_ACL);
+		break;
+	case BT_BUF_EVT:
+		net_buf_push_u8(buf, HCI_EVT);
+		break;
+	default:
+		SYS_LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
+		net_buf_unref(buf);
+		return -EINVAL;
+	}
+
+	if (buf->len > SPI_MAX_MSG_LEN) {
+		SYS_LOG_ERR("TX message too long");
+		net_buf_unref(buf);
+		return -EINVAL;
+	}
+	header_slave[STATUS_HEADER_TOREAD] = buf->len;
+
+	gpio_pin_write(gpio_dev, GPIO_IRQ_PIN, 1);
+
+	/* Coordinate transfer lock with the spi rx thread */
+	k_sem_take(&sem_spi_tx, K_FOREVER);
+	do {
+		ret = spi_transceive(spi_hci_dev, header_slave, 5,
+				     header_master, 5);
+		if (ret < 0) {
+			SYS_LOG_ERR("SPI transceive error: %d", ret);
+		}
+	} while (header_master[STATUS_HEADER_READY] != SPI_READ);
+
+	ret = spi_transceive(spi_hci_dev, buf->data, buf->len,
+			     &rxmsg, buf->len);
+	if (ret < 0) {
+		SYS_LOG_ERR("SPI transceive error: %d", ret);
+	}
+	net_buf_unref(buf);
+
+	gpio_pin_write(gpio_dev, GPIO_IRQ_PIN, 0);
+	k_sem_give(&sem_spi_rx);
+
+	return 0;
+}
+
+static void bt_tx_thread(void *p1, void *p2, void *p3)
+{
+	u8_t header_master[5];
+	u8_t header_slave[5] = { READY_NOW, SANITY_CHECK,
+				 0x00, 0x00, 0x00 };
+	struct net_buf *buf = NULL;
+	struct bt_hci_cmd_hdr cmd_hdr;
+	struct bt_hci_acl_hdr acl_hdr;
+	u8_t ret;
+
+	memset(&txmsg, 0xFF, SPI_MAX_MSG_LEN);
+
+	while (1) {
+		do {
+			ret = spi_transceive(spi_hci_dev, header_slave, 5,
+					     header_master, 5);
+			if (ret < 0) {
+				SYS_LOG_ERR("SPI transceive error: %d", ret);
+			}
+		} while ((header_master[STATUS_HEADER_READY] != SPI_READ) &&
+			 (header_master[STATUS_HEADER_READY] != SPI_WRITE));
+
+		if (header_master[STATUS_HEADER_READY] == SPI_READ) {
+			/* Unblock the spi tx thread and wait for it */
+			k_sem_give(&sem_spi_tx);
+			k_sem_take(&sem_spi_rx, K_FOREVER);
+			continue;
+		}
+
+		/* Receiving data from the SPI Host */
+		ret = spi_transceive(spi_hci_dev, &txmsg, SPI_MAX_MSG_LEN,
+				     &rxmsg, SPI_MAX_MSG_LEN);
+		if (ret < 0) {
+			SYS_LOG_ERR("SPI transceive error: %d", ret);
+		}
+
+		switch (rxmsg[PACKET_TYPE]) {
+		case HCI_CMD:
+			memcpy(&cmd_hdr, &rxmsg[1], sizeof(cmd_hdr));
+
+			buf = net_buf_alloc(&cmd_tx_pool, K_NO_WAIT);
```

Comment:
Here it is just to avoid locking up the thread (similarly how it is done with the other hci_* samples), but I'm also not completely sure what would be the best way to handle such situation here, maybe @jhedberg can give us a hand here.